lib/AST/RequirementMachine/RewriteSystem.h

```diff
@@ -225,6 +239,12 @@ class RewriteSystem final {
   /// top-level generic signature and this array is empty.
   ArrayRef<const ProtocolDecl *> Protos;
 
+  /// The requirements written in source code.
+  std::vector<StructuralRequirement> WrittenRequirements;
+
+  /// The invalid requirements.
```

Comment:
I'm worried we have too many vectors of RequirementErrors :)

This one is just redundancies detected from addExplicitRule() right? Perhaps addExplicitRule() should take a vector of RequirementErrors as an out parameter and push the errors there.

The request evaluator request would own the vector. It would first be used to collect errors from requirement desugaring. Then we can plumb it through the rule builder to collect the errors from addExplicitRule(), and then pass it down to getRedundantRequirements() to add the final set of redundancies.